docs/architecture/adr-003-dynamic-capability-store.md

```diff
@@ -0,0 +1,271 @@
+This ADR proposes such an interface & mechanism.
+
+## Decision
+
+The SDK will include a new `CapabilityKeeper` abstraction, which is responsible for provisioning, tracking, and authenticating capabilities at runtime. During application initialisation in `app.go`, the `CapabilityKeeper` will
+be hooked up to modules through unique function references (by calling `ScopeToModule`, defined below) so that it can identify the calling module when later invoked. When the initial state is loaded from disk, the `CapabilityKeeper`'s `Initialise` function will create new capability keys
+for all previously allocated capability identifiers (allocated during execution of past transactions and assigned to particular modes), and keep them in a memory-only store while the chain is running. The SDK will include a new `MemoryStore` store type, similar
+to the existing `TransientStore` but without erasure on `Commit()`, which this `CapabilityKeeper` will use to privately store capability keys.
+
+The `CapabilityKeeper` will use two stores: a regular, persistent `KVStore`, which will track what capabilities have been created by each module, and an in-memory `MemoryStore` (described below), which will
+store the actual capabilities. The `CapabilityKeeper` will define the following types & functions:
+
+The `Capability` interface, similar to `StoreKey`, provides `Name()` and `String()` methods.
+
+```golang
+type Capability interface {
+  Name() string
+  String() string
+}
+```
+
+A `CapabilityKey` is simply a struct, the address of which is taken for the actual capability.
+
+```golang
+type CapabilityKey struct {
+  name string
+}
+```
+
+A `CapabilityKeeper` contains a persistent store key, memory store key, and mapping of allocated module names.
+
+```golang
+type CapabilityKeeper struct {
+  persistentKey StoreKey
+  memoryKey MemoryStoreKey
+  moduleNames map[string]interface{}
+  sealed bool
+}
+```
+
+The `CapabilityKeeper` provides the ability to create *scoped* sub-keepers which are tied to a particular module name. These `ScopedCapabilityKeeper`s must be created at application
+initialisation and passed to modules, which can then use them to claim capabilities they receive and retrieve capabilities which they own by name, in addition
+to creating new capabilities & authenticating capabilities passed by other modules.
+
+```golang
+type ScopedCapabilityKeeper struct {
+  persistentKey StoreKey
+  memoryKey MemoryStoreKey
+  moduleName string
+}
+```
+
+`ScopeToModule` is used to create a scoped sub-keeper with a particular name, which must be unique. It MUST be called before `InitialiseAndSeal`.
+
+```golang
+func (ck CapabilityKeeper) ScopeToModule(moduleName string) ScopedCapabilityKeeper {
+  if ck.sealed {
+    panic("capability keeper is sealed")
+  }
+  if _, present := ck.moduleNames[moduleName]; present {
+    panic("cannot create multiple scoped capability keepers for the same module name")
+  }
+  ck.moduleNames[moduleName] = interface{}
```

Comment:
nit: Let this be `struct{}{}` in the implementation

Reply:
Updated.